doc/md/guides/testing/testing-procedures.mdx

```diff
@@ -0,0 +1,195 @@
+---
+id: testing-procedures
+title: Testing Stored Procedures
+slug: /guides/testing/procedures
+---
+
+Testing your database schema and migrations is crucial to ensure code behaves as expected, catch bugs early,
+and prevent regressions. Databases enforce logic, constraints, and complex relationships,
+so testing ensures these elements work correctly and remain intact after changes.
+
+In this guide we will learn how to use Atlas's [`schema test`](/testing/schema) command to test database stored
+procedures.
+
+
+## Stored Procedures
+
+Stored procedures are sets of precompiled queries grouped together to perform specific tasks and are
+stored directly on the database.
+
+:::info [Atlas Pro Feature](/features#pro-plan)
+Stored procedures are currently available only to [Atlas Pro users](/features#pro-plan). To use this feature, run:
+```
+atlas login
+```
+:::
+
+## Project Setup
+
+### Schema File
+For this example, let's assume we have the following schema, including a stored procedure:
+
+```hcl title="schema.hcl"
+schema "public" {
+}
+
+table "users" {
+  schema = schema.public
+  column "username" {
+    type = text
+  }
+  column "email" {
+    type = text
+  }
+}
+
+// highlight-start
+procedure "check_user_email" {
+  schema = schema.public
+  lang   = PLpgSQL
+  arg "username" {
+    type = character_varying
+  }
+  arg "email" {
+    type = character_varying
+  }
+  as = <<-SQL
+  BEGIN
+    IF NOT email LIKE '%@%' THEN
+      RAISE EXCEPTION 'Invalid email format';
+  END IF;
+
+  INSERT INTO users (username, email) VALUES ($1, $2);
+  END;
+  SQL
+}
+// highlight-end
+```
+
+In the schema above we have a simple `users` table and a `check_user_email` procedure. The procedure checks whether or not
+the `email` column in the `users` table has a valid email format.
+
+### Project File
+
+Before we begin testing, create a [project file](/atlas-schema/projects#project-files) named
+`atlas.hcl`.
+
+In this file we will create an environment, specify the source of our schema,
+and a URL for our [dev database](/concepts/dev-database).
+
+We will also create a file named `schema.test.hcl` to write our tests, and
+add it to the `atlas.hcl` file in the test block.
+
+```hcl title="atlas.hcl"
+env "dev" {
+  src = "file://schema.hcl"
+  dev = "docker://postgres/15/dev?search_path=public"
+
+  # Test configuration for local development.
+  test {
+    schema {
+      src = ["schema.test.hcl"]
+    }
+  }
+}
+```
+
+## Writing Tests
+
+### Simple Test
+
+Let's start off with a simple test that will check that the stored procedures correctly recognizes valid emails.
+
+```hcl title="schema.test.hcl"
+test "schema" "procedure" {
+  # Attempt to create user with invalid email (missing @)
+  catch {
+    sql = "CALL check_user_email('new_user', 'invalidemail')"
+  }
+  log {
+    message = "Invalid email passed"
+  }
```

Comment:
better to add another `exec` block that query the affects and compare the result.